### DISCUSSION TITLE
SQL Relationship (Many)

### DISCUSSION THREAD

**Discussion by AndyGUK** [Budibase/budibase#3490] [Help] [answered]
I cannot establish a relationship between my tables. I have three tables:
![image](https://user-images.githubusercontent.com/73335367/143017900-442a517d-c3b6-4556-943b-722deea81df1.png)

My setup looks like this:
![image](https://user-images.githubusercontent.com/73335367/143017995-ac3ae743-d325-4cf7-9e96-3a2236b08476.png)

Any my display columns look like this:
![image](https://user-images.githubusercontent.com/73335367/143018086-5a63eb47-fa91-490d-8749-e6d6ec9bec98.png)

However, when I navigate to the form I do not get a list of all the 'franchises' via the options selector:
![image](https://user-images.githubusercontent.com/73335367/143018194-b81aa73f-7516-4e7b-b7b9-cf77d7068ce9.png)

What am I doing wrong? Also, there doesn't seem to be a lot of detailed documentation for this considering it's a low-code app?!

**Comment by mike12345567** [comment, marked as answer]
Hi @AndyGUK - from your ER diagram it looks like you might not be able to represent this currently in Budibase, this is something we'll look at soon.

The problem is that right now Budibase assumes that your rows will be related on the primary keys of the two tables, with the foreign keys being defined on the junction table; in your case, that would mean that `franchise_id` and `idtickets` (on the `franchise_ticket_logs`) would be the foreign keys targeting the primary keys of `franchise` and `tickets` respectively. You haven't labelled the primary key on your `franchise` table, but I assume it is the `uid` field referenced at the top of it - Budibase would expect this field to be the one related on. You can see an example of a many to many relationship in Budibase here: https://github.com/Budibase/budibase/blob/develop/packages/server/scripts/integrations/postgres/init.sql.

The next version of Budibase will also be able to create/edit tables in SQL, as well as creating/updating the keys/columns, which should help with a lot of this.

There are some changes coming in this area to provide better error feedback around table fetching and creating relationships etc in SQL databases - its a relatively new feature so we're still building it out.

**Reply by AndyGUK** [reply]
Thanks @mike12345567, this is really helpful to know. 

Despite the ER diagram, is it still possible to add a field to my form which allows users to select a franchise, possibly passing the franchise_id also?

If not, what's the best approach to creating this sort of functionality?

**Reply by mike12345567** [reply]
You can customise a form and use a data provider to do this right now, so what you would do is:

1. Setup a data provider, which fetches data from your `franchise` table
2. Create a form within the data provider, set the form schema to your `tickets` table
3. Create form, add a field group to it
4. Add an options picker to the field group
5. In your options picker simply type the name of the field you want to set the franchise ID in, then set it up to use the data provider as its source of options, like so:
![image](https://user-images.githubusercontent.com/4407001/143050322-61cd3957-69a4-40b8-a4cf-4b2cbb459bbe.png)

If you want to generate a ticket and a link in the `franchise_ticket_log` junction table at the same time you'll need to setup a custom query, which does both of the operations in SQL and takes the franchise ID as a parameter to the custom query.

This will be a lot easier once we make it possible to specify all of the fields involved in the relationship, but we just need to do a little bit of UI work there as its getting pretty complicated already!

**Reply by AndyGUK** [reply]
Hmmm.... I seem to have copied this word for word yet it's not working. Strangely, I'm not actually seeing the `franchise` data being pulled through to the GUI, could this be why it's not working?

**Reply by mike12345567** [reply]
Yes that could definitely be an issue, if you aren't able to see any `franchise` data coming back then its possible theres something wrong with the config.

One thing to note (this is much more clear in the next version of Budibase) is that you need a primary key defined for every table you fetch data from, if one is missing then you might have a scenario like this.

Alternatively you can try using custom queries to just select the data from your table and display it through a data provider, even just to see if that works.

**Reply by AndyGUK** [reply]
Ah, I was missing the PK. Thanks for your help, the form/ data seems to be worked as intended now.

Cheers for your help!

**Reply by mike12345567** [reply]
Glad you could resolve it! In the next version of Budibase this will be very obvious, we've added a lot of various error scenarios as things we can catch and provide information to the user about, as you can see in this PR: https://github.com/Budibase/budibase/pull/3444
